Content.Server/Automod/AutomodManager.cs

```diff
@@ -0,0 +1,243 @@
+            return true;
+
+        foreach (var filter in automodFilters)
+        {
+            var regexMatches = filter.Regex.Matches(inputText);
+            if (regexMatches.Count == 0)
+                continue;
+
+            var textMatches = new List<(string, int)>();
+            foreach (Match match in regexMatches)
+            {
+                var str = match.ToString();
+                textMatches.Add((str, match.Index));
+            }
+
+            if (!_protoMan.TryIndex<AutomodActionGroupPrototype>(filter.ActionGroup, out var censorGroup))
+            {
+                _log.Error($"AutomodActionGroupPrototype \"{filter.ActionGroup}\" not found.");
+                continue;
+            }
+
+            var skip = false;
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                if (!censorAction.Skip(inputText, textMatches))
+                    continue;
+
+                skip = true;
+                break;
+            }
+            if (skip)
+                continue;
+
+            var displayName = GetDisplayName(filter);
+
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                passes &= censorAction.RunAction(session, inputText, textMatches, filter, displayName, _entMan);
+            }
+        }
+
+        return passes;
+    }
+
+    private string GetDisplayName(AutomodFilterDef filter)
+    {
+        return filter.Id != null ? $"{filter.DisplayName} (automod#{filter.Id})" : $"{filter.DisplayName} (automod#na)";
+    }
+
+    #endregion
+
+    #region Reload
+
+    public async void ReloadAutomodFilters()
+    {
+        _regexFilters.Clear();
+
+        var filters = await _db.GetAllAutomodFiltersAsync();
+
+        foreach (var filter in filters)
+        {
+            AddFilter(filter);
+        }
+    }
+
+    #endregion
+
+    #region Create
+
+    public async void CreateFilter(string pattern,
+        AutomodFilterType filterType,
+        string actionGroup,
+        AutomodTarget targets,
+        string name)
+    {
+        CreateFilter(new AutomodFilterDef(pattern, filterType, actionGroup, targets, name));
+    }
+
+    public async void CreateFilter(AutomodFilterDef automod)
+    {
+        automod = await _db.AddAutomodFilterAsync(automod);
+        AddFilter(automod);
+    }
+
+    /// <summary>
+    /// Adds a filter to the manager lists.
+    /// </summary>
+    /// <param name="automod"></param>
+    private void AddFilter(AutomodFilterDef automod)
+    {
+        foreach (var targetFlag in Enum.GetValues<AutomodTarget>())
+        {
+            if (targetFlag == AutomodTarget.None)
+                continue;
+            if (!automod.TargetFlags.HasFlag(targetFlag))
+                continue;
+
+            if (automod.FilterType == AutomodFilterType.Regex)
+            {
+                var list = _regexFilters.GetOrNew(targetFlag);
+                _regexFilters.TryAdd(targetFlag, list);
```

Comment:
This `TryAdd` isn't necessary, `GetOrNew()` already does that.

Reply:
Removed the TryAdd [18d9268](https://github.com/space-wizards/space-station-14/pull/28570/commits/18d9268258103659f02311a1cd8abc23eceaa6d7)